### PR TITLE
feat: Add L2 regularization (weight decay) support

### LINKED ISSUE #21: [Core] L2 Regularization (Weight Decay)
### Description
We currently have no mechanism to prevent overfitting. L2 regularization is the most effective way to address this in small MLPs.

### Implementation Details
- Add a `weight_decay` (lambda) parameter to `optimizer.rs`.
- In `Linear::backward`, add `lambda * weight` to the `grad_weights` before performing the update.
- Expose this parameter through `lib.rs` to the Python API.

### Acceptance Criteria
- Higher weight decay values lead to smaller weight magnitudes in the saved JSON model.

### PR DESCRIPTION
**Pull request by verdhanyash**
## Summary
Implements L2 regularization (weight decay) to help prevent overfitting in small MLPs.

Closes #21

### Rust Core
- **`optimizer.rs`**: Added `weight_decay` field to SGD optimizer with `with_weight_decay()` constructor
- **`layers.rs`**: Modified `Linear::update()` to apply L2 penalty: `grad_total = grad_loss + λ * weight`
- **`model.rs`**: Updated `train()` to accept and pass `weight_decay` parameter
- **`lib.rs`**: Exposed `weight_decay` to Python API with default value `0.0`

### Python API
- **`api.py`**: Added `weight_decay` parameter to `Model.train()` with documentation
- **Bonus fix**: `predict()` now works without arguments after training (caches training data)

## Usage
```python
from etna import Model

model = Model("data.csv", target="label")

# Without regularization (default, backward compatible)
model.train(epochs=100, lr=0.01)

# With L2 regularization
model.train(epochs=100, lr=0.01, weight_decay=0.001)
```

## ScreenShot 
Screenshot 2026-01-06 004635

### CODE REVIEW

**Review by Aamod007 — APPROVED** [review]
## Review: Approve

Hi @verdhanyash, thank you for the contribution.

Build passed, all 24 tests passed. L2 regularization is correctly implemented with the penalty applied in `Linear::update()` and biases excluded per standard practice. Python API is properly exposed with backward-compatible defaults.

Ready to merge after further reviews.

---

Regards,
**Aamod Kumar**
Mentor, SWOC '26

**Comment by verdhanyash** [comment]
Hi @Aamod007 , thank you for the review and approval. Glad to hear the build passed and the L2 implementation meets standard practice.
Thanks for the guidance.
Regards,
verdhanyash




> ## Review: Approve
> Hi @verdhanyash, thank you for the contribution.
> 
> Build passed, all 24 tests passed. L2 regularization is correctly implemented with the penalty applied in `Linear::update()` and biases excluded per standard practice. Python API is properly exposed with backward-compatible defaults.
> 
> Ready to merge after further reviews.
> 
> Regards, **Aamod Kumar** Mentor, SWOC '26

**Comment by verdhanyash** [comment]
Hi @debug-soham , thank you for the appreciation and for assigning Issue #21 so the contribution can be counted. Apologies for opening the PR before the official assignment — I’ve noted the SWoC process and will ensure I get assigned before opening any future PRs.

Thanks for the guidance!
Regards,
verdhanyash




> @verdhanyash Great work on the L2 implementation! However, please remember that SWoC rules require an official assignment before opening a PR. Since Issue #21 is currently unassigned, I’ll assign it to you so this contribution counts. Please ensure you follow the assignment process for all future tasks.
